api/integrations/github/views.py

```diff
@@ -0,0 +1,166 @@
+from functools import wraps
+
+import requests
+from django.conf import settings
+from django.http import JsonResponse
+from rest_framework import status, viewsets
+from rest_framework.decorators import api_view, permission_classes
+from rest_framework.permissions import IsAuthenticated
+from rest_framework.response import Response
+
+from integrations.github.client import generate_token
+from integrations.github.constants import GITHUB_API_URL, GITHUB_API_VERSION
+from integrations.github.models import GithubConfiguration, GithubRepository
+from integrations.github.permissions import HasPermissionToGithubConfiguration
+from integrations.github.serializers import (
+    GithubConfigurationSerializer,
+    GithubRepositorySerializer,
+    RepoQuerySerializer,
+)
+from organisations.models import Organisation
+
+
+def github_auth_required(func):
+
+    @wraps(func)
+    def wrapper(request, organisation_pk):
+
+        if not GithubConfiguration.has_github_configuration(
+            organisation_id=organisation_pk
+        ):
+            return Response(
+                data={
+                    "detail": "This Organisation doesn't have a valid GitHub Configuration"
+                },
+                content_type="application/json",
+                status=status.HTTP_400_BAD_REQUEST,
+            )
+        return func(request, organisation_pk)
+
+    return wrapper
+
+
+class GithubConfigurationViewSet(viewsets.ModelViewSet):
+    permission_classes = (IsAuthenticated, HasPermissionToGithubConfiguration)
+    serializer_class = GithubConfigurationSerializer
+    model_class = GithubConfiguration
+
+    def perform_create(self, serializer):
+        organisation_id = self.kwargs["organisation_pk"]
+        serializer.save(organisation_id=organisation_id)
+
+    def get_queryset(self):
+        return GithubConfiguration.objects.filter(
+            organisation_id=self.kwargs["organisation_pk"]
+        )
+
+
+class GithubRepositoryViewSet(viewsets.ModelViewSet):
+    permission_classes = (IsAuthenticated, HasPermissionToGithubConfiguration)
+    serializer_class = GithubRepositorySerializer
+    model_class = GithubRepository
+
+    def perform_create(self, serializer):
+        github_configuration_id = self.kwargs["github_pk"]
+        serializer.save(github_configuration_id=github_configuration_id)
+
+    def get_queryset(self):
+        return GithubRepository.objects.filter(
+            github_configuration=self.kwargs["github_pk"]
+        )
+
+
+@api_view(["GET"])
+@permission_classes([IsAuthenticated, HasPermissionToGithubConfiguration])
+@github_auth_required
```

Comment:
Is there a reason why we are not using class based views here?

Reply:
Yes, it is because this view will have only `GET ` method always and I have seen that we use the `@api_view` decorator in several other similar views.